g3/main.py

```diff
@@ -0,0 +1,77 @@
+from enum import Enum
+from typing import Annotated
+
+import typer
+
+app = typer.Typer()
+
+
+class MessageTone(str, Enum):
+    PROFESSIONAL = "professional"
+    PERSONAL = "personal"
+    FRIENDLY = "friendly"
+    FUNNY = "funny"
+
+
+@app.command()
+def commit(
+    tone: Annotated[
+        MessageTone,
+        typer.Option("--tone", "-t", help="The tone of voice that will be used to generate the commit message"),
+    ] = MessageTone.FRIENDLY,
+    jira: Annotated[
+        str, typer.Option("--jira", "-j", help="A Jira ticket number that should be referenced in the commit message")
+    ] = "",
+    include: Annotated[
+        str,
+        typer.Option("--include", "-i", help="Text content that should be appended to the commit message"),
+    ] = "",
+    edit: Annotated[
+        str,
+        typer.Option(
+            "--edit", "-e", help="A commit hash that points to a commit which will be used instead of staged changes"
+        ),
+    ] = "",
+) -> None:
+    """Generate a new commit message for staged changes"""
+    typer.echo(f"Generating commit message with {tone.value} tone.")
+    if edit:
+        typer.echo(f"For existing commit: {edit}")
+    if jira:
+        typer.echo(f"Referencing Jira ticket: {jira}")
+    if include:
+        typer.echo(f"Including additional text:\n{include}")
+
+
+@app.command()
+def pr(
+    tone: Annotated[
+        MessageTone,
+        typer.Option("--tone", "-t", help="The tone of voice that will be used to generate the PR description"),
+    ] = MessageTone.FRIENDLY,
+    jira: Annotated[
+        str, typer.Option("--jira", "-j", help="A Jira ticket number that should be referenced in the PR description")
+    ] = "",
+    include: Annotated[
+        str,
+        typer.Option("--include", "-i", help="Text content that should be appended to the PR description"),
+    ] = "",
+    edit: Annotated[
+        str,
+        typer.Option(
+            "--edit", "-e", help="A hash that points to a previous PR which will be used instead of the current branch"
+        ),
+    ] = "",
+) -> None:
+    """Generate a new pull request description for the current branch"""
+    typer.echo(f"Generating PR description with {tone.value} tone.")
+    if edit:
+        typer.echo(f"For previous PR: {edit}")
+    if jira:
+        typer.echo(f"Referencing Jira ticket: {jira}")
+    if include:
+        typer.echo(f"Including additional text:\n{include}")
```

Comment:
I am a bit concerned about how these can work along with `git` and `gh` CLIs so that we end up providing an all-in-one CLI. 

## Approach 1
* Support `g commit --title .. --body` so that we support manual commit messages too.
* Support `g pr --title .. --body ..` so that we support manual PR messages too.
* Add a fallback command (that won't be documented though by typer) so that we delegate to `git` otherwise as follows:

```
@app.callback(invoke_without_command=True)
def git(*args):
    subprocess.run(["git", *args], check=True, stdout=subprocess.PIPE).stdout
```

## Approach 2
Keep this CLI separate and work for the [facade](https://docs.google.com/document/d/1b3SkLVnrB_P-lYsD2Kv-fO-oUHciovQbTQbKsNswHtA/edit#heading=h.awqexfhtx1tf) API in another module that will call this in case no message is provided by the users. 

In this case, we should make some space in `main.py` for the delegation to the appropriate module moving these 2 commands to a separate `g3/generate` package.

This approach has the drawback, though, that we will lose the nice docs of typer.

## Approach 3
Drop `git` and `gh` support for the first version of the hackathon, since this is an AI-focused competition, and work on that in the 2nd version.

WDYT?

Reply:
My vote goes to option 3. I think that in the scope of the hackathon it would be a waste of time to try to make a solution that handles `git` and `gh` transparently. If we expand the project in the future, we can do a proper brainstorming about it, without the time constraint of a few hours.